group/edwards25519/point.go

```diff
@@ -290,3 +294,240 @@ func (P *point) IsCanonical(s []byte) bool {
+	}
+
+	padDom, err := i2OSP(len(domainSeparator), 1)
+	if err != nil {
+		return nil, err
+	}
+
+	dstPrime := append([]byte(domainSeparator), padDom...)
+	byteLenStr, _ := i2OSP(byteLen, 2)
+	zeroPad, _ := i2OSP(0, 1)
+	zPad, _ := i2OSP(0, h.BlockSize())
+
+	// mPrime = Z_pad || msg || l_i_b_str || I2OSP(0, 1) || DST_prim
+	mPrime := make([]byte, 0, len(zPad)+len(m)+len(byteLenStr)+len(zeroPad)+len(dstPrime))
+	mPrime = append(mPrime, zPad...)
+	mPrime = append(mPrime, m...)
+	mPrime = append(mPrime, byteLenStr...)
+	mPrime = append(mPrime, zeroPad...)
+	mPrime = append(mPrime, dstPrime...)
+
+	// b0 = H(msg_prime)
+	h.Reset()
+	h.Write([]byte(mPrime))
+	b0 := h.Sum(nil)
+
+	// b_1 = H(b_0 || I2OSP(1, 1) || DST_prime
+	h.Reset()
+	h.Write(b0)
+	onePad, _ := i2OSP(1, 1)
+	h.Write([]byte(onePad))
+	h.Write([]byte(dstPrime))
+	b1 := h.Sum(nil)
+
+	bFinal := make([]byte, 0, len(b1)*(ell+1))
+	bFinal = append(bFinal, b1...)
+	bPred := b1
+	for i := 2; i <= ell; i++ {
+		x, err := byteXor(bPred, b0, bPred)
+		if err != nil {
+			return nil, err
+		}
+		ithPad, _ := i2OSP(i, 1)
+
+		h.Reset()
+		h.Write(x)
+		h.Write(ithPad)
+		h.Write(dstPrime)
+
+		bPred = h.Sum(nil)
+		bFinal = append(bFinal, bPred...)
+	}
+
+	return bFinal[:byteLen], nil
+}
+
+func i2OSP(x int, xLen int) ([]byte, error) {
+	b := big.NewInt(int64(x))
+	s := b.Bytes()
+	if len(s) > xLen {
+		return nil, xerrors.Errorf("input %d superior to max length %d", len(s), xLen)
+	}
+
+	pad := make([]byte, (xLen - len(s)))
+	return append(pad, s...), nil
+}
+
+func byteXor(dst, b1, b2 []byte) ([]byte, error) {
+	if !(len(dst) == len(b1) && len(b2) == len(b1)) {
+		return nil, xerrors.New("incompatible lengths")
+	}
+
+	for i := 0; i < len(dst); i++ {
+		dst[i] = b1[i] ^ b2[i]
+	}
+
+	return dst, nil
+}
+
+func curve25519Elligator2(u fieldElement) (xn, xd, yn, yd fieldElement) {
+	// Some const needed
+	var one fieldElement
+	feOne(&one)
+
+	j := fieldElement{486662, 0, 0, 0, 0, 0, 0, 0, 0, 0}
+
+	// c1 = (q + 3) / 8
+	// c2 = 2^c1
+	// c3 = sqrt(-1)
+	// c4 = (q - 5) / 8
+	// Computed with sagemath
+	c2 := fieldElement{34513073, 25610706, 9377949, 3500415, 12389472, 33281959, 41962654, 31548777, 326685, 11406482}
+	c3 := fieldElement{34513072, 25610706, 9377949, 3500415, 12389472, 33281959, 41962654, 31548777, 326685, 11406482}
+	c4, _ := new(big.Int).SetString("7237005577332262213973186563042994240829374041602535252466099000494570602493", 10)
+
+	// Temporary variables
+	var tv1, tv2, tv3, x1n, gxd, gx1, gx2 fieldElement
+	var y, y1, y2, y11, y12, y21, y22, x2n fieldElement
+	var e1, e2, e3, e4 int32
+
+	feSquare2(&tv1, &u)
+
+	feAdd(&xd, &one, &tv1)
+	feNeg(&x1n, &j)
+	feSquare(&tv2, &xd)
+	feMul(&gxd, &tv2, &xd)
+	feMul(&gx1, &j, &tv1)
+	feMul(&gx1, &gx1, &x1n)
+	feAdd(&gx1, &gx1, &tv2)
+	feMul(&gx1, &gx1, &x1n)
+	feSquare(&tv3, &gxd)
+	feSquare(&tv2, &tv3)
+	feMul(&tv3, &tv3, &gxd)
+	feMul(&tv3, &tv3, &gx1)
+	feMul(&tv2, &tv2, &tv3)
+
+	// compute y11 = tv2 ^ c4
```

Comment:
like this, but for each step

Reply:
for `curve25519Elligator2` I used: https://datatracker.ietf.org/doc/html/rfc9380#name-curve25519-q-5-mod-8-k-1
That was put in the function doc for clarity.